File: packages/core/src/storage-digest.ts
import type { DurableSeriesState, TaskEvent } from './types.js'

const encoder = new TextEncoder()

/**
 * Minimal JSON canonicalization used by the hot/cold storage protocol.
 *
 * Objects are recursively key-sorted and numbers are normalized by numeric
 * value, so digests do not depend on JavaScript object insertion order or on
 * insignificant JSON number formatting.
 */
export function canonicalJson(value: unknown): string {
  if (value === null) return 'null'

  switch (typeof value) {
    case 'boolean':
      return value ? 'true' : 'false'
    case 'number':
      if (!Number.isFinite(value)) throw new TypeError('Archive digest JSON numbers must be finite')
      return String(value)
    case 'string':
      return JSON.stringify(value)
    case 'object':
      if (Array.isArray(value)) {
        const encoded: string[] = []
        for (let index = 0; index < value.length; index++) {
          if (!Object.hasOwn(value, index)) {
            throw new TypeError('Archive digest arrays must not contain holes')
          }
          encoded.push(canonicalJson(value[index]))
        }
        return `[${encoded.join(',')}]`
      }
      if (
        (Object.getPrototypeOf(value) !== Object.prototype &&
          Object.getPrototypeOf(value) !== null) ||
        ('toJSON' in value && typeof value.toJSON === 'function')
      ) {
        throw new TypeError('Archive digest objects must be plain JSON objects')
      }
      return `{${Object.keys(value as Record<string, unknown>)
        .filter((key) => (value as Record<string, unknown>)[key] !== undefined)
        .sort(compareUtf8)
        .map(
          (key) =>
            `${JSON.stringify(key)}:${canonicalJson((value as Record<string, unknown>)[key])}`,
        )
        .join(',')}}`
    default:
      throw new TypeError(`Archive digest cannot encode ${typeof value}`)
  }
}

export function archiveEventRecord(event: TaskEvent): string {
  if (!Number.isSafeInteger(event.index) || event.index < 0) {
    throw new TypeError('Archive digest event index must be a non-negative safe integer')
  }
  if (!Number.isFinite(event.timestamp)) {
    throw new TypeError('Archive digest event timestamp must be finite')
  }
  return canonicalJson([
    'taskcast-event-v1',
    event.id,
    event.taskId,
    String(event.index),
    String(event.timestamp),
    event.type,
    event.level,
    event.data,
    event.seriesId ?? null,
    event.seriesMode ?? null,
    event.seriesAccField ?? null,
  ])
}

export function durableSeriesStateRecord(state: DurableSeriesState): string {
  return canonicalJson([
    'taskcast-series-v1',
    state.taskId,
    state.seriesId,
    state.mode,
    String(state.throughIndex),
    archiveEventRecord(state.event),
  ])
}

export async function sha256Hex(value: string): Promise<string> {
  const digest = await globalThis.crypto.subtle.digest('SHA-256', encoder.encode(value))
  return Array.from(new Uint8Array(digest), (byte) => byte.toString(16).padStart(2, '0')).join('')
}

export async function computeArchiveBatchDigest(
  previousBatchDigest: string | null,
  events: readonly TaskEvent[],
  seriesLatest: readonly DurableSeriesState[],
): Promise<string> {
  return sha256Hex(
    [
      'taskcast-batch-v1',
      previousBatchDigest ?? '',
      ...events.map(archiveEventRecord),
      ...sortedSeriesRecords(seriesLatest),
    ].join('\n'),
  )
}

export async function computeArchiveSourcePageDigest(
  events: readonly TaskEvent[],
): Promise<string> {
  return sha256Hex(
    [
      'taskcast-source-page-v1',
      ...events.map((event) => canonicalJson([String(event.index), event.id])),
    ].join('\n'),
  )
}

export async function computeArchiveSourceDigest(
  pageDigests: readonly string[],
): Promise<string> {
  return sha256Hex(['taskcast-source-v1', ...pageDigests].join('\n'))
}

export async function computeSeriesStateDigest(
  seriesLatest: readonly DurableSeriesState[],
): Promise<string> {
  return sha256Hex(['taskcast-series-state-v1', ...sortedSeriesRecords(seriesLatest)].join('\n'))
}

function sortedSeriesRecords(seriesLatest: readonly DurableSeriesState[]): string[] {
  return [...seriesLatest]
    .sort((left, right) => {
      const taskOrder = compareUtf8(left.taskId, right.taskId)
      return taskOrder !== 0 ? taskOrder : compareUtf8(left.seriesId, right.seriesId)
    })
    .map(durableSeriesStateRecord)
}

function compareUtf8(left: string, right: string): number {
  const leftBytes = encoder.encode(left)
  const rightBytes = encoder.encode(right)
  const length = Math.min(leftBytes.length, rightBytes.length)
  for (let index = 0; index < length; index++) {
    const difference = leftBytes[index]! - rightBytes[index]!
    if (difference !== 0) return difference
  }
  return leftBytes.length - rightBytes.length
}

File: rust/taskcast-core/tests/storage_digest.rs
use taskcast_core::archive::{
    archive_event_record, canonical_json, compute_archive_batch_digest,
    compute_archive_source_digest, compute_archive_source_page_digest, compute_series_state_digest,
};
use taskcast_core::types::{DurableSeriesState, Level, SeriesMode, TaskEvent};

fn fixture_event() -> TaskEvent {
    TaskEvent {
        id: "evt-7".to_string(),
        task_id: "task-1".to_string(),
        index: 7,
        timestamp: 1_700_000_000_123.0,
        r#type: "llm.delta".to_string(),
        level: Level::Info,
        data: serde_json::json!({
            "z": [3, { "b": true, "a": null }],
            "a": "hello"
        }),
        series_id: Some("output".to_string()),
        series_mode: Some(SeriesMode::Accumulate),
        series_acc_field: Some("delta".to_string()),
        series_snapshot: None,
        _accumulated_data: None,
    }
}

#[test]
fn canonical_json_is_independent_of_object_insertion_order() {
    assert_eq!(
        canonical_json(&serde_json::json!({ "z": 1, "a": { "y": 2, "b": 3 } })).unwrap(),
        r#"{"a":{"b":3,"y":2},"z":1}"#
    );
    assert_eq!(
        canonical_json(&serde_json::json!([
            1e21,
            1e20,
            1e-7,
            1e-6,
            -0.0,
            u64::MAX,
            667082108456853.2_f64
        ]))
        .unwrap(),
        "[1e+21,100000000000000000000,1e-7,0.000001,0,18446744073709552000,667082108456853.2]"
    );
}

#[test]
fn digest_protocol_matches_typescript_fixture() {
    let event = fixture_event();
    let series = DurableSeriesState {
        task_id: "task-1".to_string(),
        series_id: "output".to_string(),
        mode: SeriesMode::Accumulate,
        event: event.clone(),
        through_index: 7,
    };

    assert_eq!(
        archive_event_record(&event).unwrap(),
        r#"["taskcast-event-v1","evt-7","task-1","7","1700000000123","llm.delta","info",{"a":"hello","z":[3,{"a":null,"b":true}]},"output","accumulate","delta"]"#
    );
    assert_eq!(
        compute_archive_batch_digest(None, &[event.clone()], &[series.clone()]).unwrap(),
        "fcaa595fb88f042f2e86decfa48dd46483f80bd7edb04d5c8b7a5876345003d8"
    );
    let page_digest = compute_archive_source_page_digest(&[event]).unwrap();
    assert_eq!(
        page_digest,
        "a494e9437592b3a58deb02a98e414ba87cd591079695bb2ebd4dd4c04d506fc8"
    );
    assert_eq!(
        compute_archive_source_digest(&[page_digest]),
        "d25d5e5dd7d8dba54b03d2bf56156593d5d8394ecbf0a65ce1e3eae8fe1c050a"
    );
    assert_eq!(
        compute_series_state_digest(&[series]).unwrap(),
        "5b58f61debb90051f9a92459b3eb98776cfa02241d7cd9c09fedfa863c70d750"
    );
}
